src/SALib/sample/sobol.py

```diff
@@ -0,0 +1,234 @@
+        If `seed` is None the `numpy.random.Generator` generator is used.
+        If `seed` is an int, a new ``Generator`` instance is used,
+        seeded with `seed`.
+        If `seed` is already a ``Generator`` instance then that instance is
+        used. Default is None.
+
+    References
+    ----------
+    .. [1] Campolongo, F., Saltelli, A., Cariboni, J., 2011.
+           From screening to quantitative sensitivity analysis.
+           A unified approach.
+           Computer Physics Communications 182, 978–988.
+           https://doi.org/10.1016/j.cpc.2010.12.039
+
+    .. [2] Owen, A. B., 2020.
+           On dropping the first Sobol' point.
+           arXiv:2008.08051 [cs, math, stat].
+           Available at: http://arxiv.org/abs/2008.08051
+           (Accessed: 20 April 2021).
+
+    .. [3] Saltelli, A., 2002.
+           Making best use of model evaluations to compute sensitivity indices.
+           Computer Physics Communications 145, 280–297.
+           https://doi.org/10.1016/S0010-4655(02)00280-1
+
+    .. [4] Sobol', I.M., 2001.
+           Global sensitivity indices for nonlinear mathematical models and
+           their Monte Carlo estimates.
+           Mathematics and Computers in Simulation,
+           The Second IMACS Seminar on Monte Carlo Methods 55, 271–280.
+           https://doi.org/10.1016/S0378-4754(00)00270-6
+    """
+    D = problem['num_vars']
+    groups = _check_groups(problem)
+
+    # Create base sequence - could be any type of sampling
+    qrng = qmc.Sobol(d=2*D, scramble=scramble, seed=seed)
+
+    # fast-forward logic
+    if skip_values > 0:
+        M = skip_values
+        if not ((M & (M-1) == 0) and (M != 0 and M-1 != 0)):
+            msg = f"""
+            Convergence properties of the Sobol' sequence is only valid if
+            `skip_values` ({M}) is a power of 2.
+            """
+            warnings.warn(msg, stacklevel=2)
+
+        # warning when N > skip_values
+        # see https://github.com/scipy/scipy/pull/10844#issuecomment-673029539
+        n_exp = int(np.log(N, 2))
+        m_exp = int(np.log(M, 2))
+        if n_exp > m_exp:
+            msg = (
+                "Convergence may not be valid as the number of "
+                "requested samples is"
+                f" > `skip_values` ({N} > {M})."
+            )
+            warnings.warn(msg, stacklevel=2)
+
+        qrng.fast_forward(M)
+    elif skip_values < 0 or isinstance(skip_values, int):
+        raise ValueError("`skip_values` must be a positive integer.")
+
+    # sample Sobol' sequence
+    base_sequence = qrng.random(N)
+
+    if not groups:
+        Dg = problem['num_vars']
+    else:
+        G, group_names = compute_groups_matrix(groups)
+        Dg = len(set(group_names))
+
+    if calc_second_order:
+        saltelli_sequence = np.zeros([(2 * Dg + 2) * N, D])
+    else:
+        saltelli_sequence = np.zeros([(Dg + 2) * N, D])
+    index = 0
+
+    for i in range(N):
```

Comment:
Not completely necessary but I suggest we replace these nested loops with `itertools.product()`

Reply:
I am not sure how to go about this. The loops need to update `index` in the first level. Maybe this can be kept for latter. And actually there might be another way to write these with NumPy operations.